# Assignment 3 (loops and conditions)
#### 140 points

__(Instruction)__ The purpose of this assignment is to practice using for and while loops and if-elif-else conditions

* Although there are functions available in Python to do most of the problems in this assignment, it is important that we know how to solve these problems using loops and conditions. This way, we can handle more complex work that cannot be done by pre-existing functions. In addition, we can also make our own functions by using loops and conditions. In such case, we are not limited by the resources online.

Due date: October 22, 2020, 11:59pm

### Question 1
Write a program to calculate the factorial of a positive integer input by the user. Recall that the factorial function is given by $x! = x (x-1) (x-2) ... (2) (1)$ so that $1!=1$, $2!=2$, $3!=6$, $4!=24$, ...

(a) Write the factorial function using a Python `while` loop. 
* (5 points)

In [4]:
print('Please provide an integer to calculate its factorial')
x = int(input())
fact = x-1
product = x
while fact > 1:
    product = product * fact
    fact = fact - 1
print('%i ! = %i'%(x,product))

Please provide an integer to calculate its factorial


 5


5 ! = 120


(b) Write the factorial function using a Python `for` loop. 
*(5 points)

In [7]:
print('Please provide an integer to calculate its factorial')
n = int(input())
product = n
for i in range(n,2,-1):
    product = product*(i-1)
print('%i ! = %i'%(n,product))

Please provide an integer to calculate its factorial


 4


4 ! = 24


(c) With a given integral number n, write a program to generate a dictionary that contains (i, factorial of i) such that is an integral number between 1 and n (both included) 

Suppose the following input is supplied to the program:
8

Then, the output should be:
{1: 1, 2: 2, 3: 6, 4: 24, 5: 120, 6: 720, 7: 5040, 8: 40320}
* (10 points)

In [13]:
print('Please provide an integer greater than or equal to 1')
n = int(input())
dict = {}
for i in range(1,n+1):
    if i >1:
        dict.update({"%i"%i:"%i"%(i*(int(dict.get('%i'%(i-1)))))})
    else:
        dict.update({"%i"%i:'1'})
        
dict

Please provide an integer greater than or equal to 1


 6


{'1': '1', '2': '2', '3': '6', '4': '24', '5': '120', '6': '720'}

### Question 2
Write a program which will find all such numbers which are divisible by 7 but are not a multiple of 5, between 2000 and 3200 (both included).
* (5 points)

In [17]:
import numpy as np
nums_by_7_ex_5 = []
for i in range(2000,3201):
    if i%7 == 0 and i%5 != 0:
        nums_by_7_ex_5.append(i)
nums_by_7_ex_5 = np.asarray(nums_by_7_ex_5)
nums_by_7_ex_5

        

array([2002, 2009, 2016, 2023, 2037, 2044, 2051, 2058, 2072, 2079, 2086,
       2093, 2107, 2114, 2121, 2128, 2142, 2149, 2156, 2163, 2177, 2184,
       2191, 2198, 2212, 2219, 2226, 2233, 2247, 2254, 2261, 2268, 2282,
       2289, 2296, 2303, 2317, 2324, 2331, 2338, 2352, 2359, 2366, 2373,
       2387, 2394, 2401, 2408, 2422, 2429, 2436, 2443, 2457, 2464, 2471,
       2478, 2492, 2499, 2506, 2513, 2527, 2534, 2541, 2548, 2562, 2569,
       2576, 2583, 2597, 2604, 2611, 2618, 2632, 2639, 2646, 2653, 2667,
       2674, 2681, 2688, 2702, 2709, 2716, 2723, 2737, 2744, 2751, 2758,
       2772, 2779, 2786, 2793, 2807, 2814, 2821, 2828, 2842, 2849, 2856,
       2863, 2877, 2884, 2891, 2898, 2912, 2919, 2926, 2933, 2947, 2954,
       2961, 2968, 2982, 2989, 2996, 3003, 3017, 3024, 3031, 3038, 3052,
       3059, 3066, 3073, 3087, 3094, 3101, 3108, 3122, 3129, 3136, 3143,
       3157, 3164, 3171, 3178, 3192, 3199])

### Question 3
(a) Consider the matrix list `x = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]`. Write a list comprehension to extract the last column of the matrix \[3, 6, 9\].
* (5 points)

In [18]:
x = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
last = [i[-1] for i in x]
last

[3, 6, 9]

[8, 50, 128]

(b) Write another list comprehension to create a vector of twice the square of the middle column `[8, 50, 128]`.
* (5 points)

In [51]:
twice_sq_mid = [2*j[1]**2 for j in x]
twice_sq_mid

[8, 50, 128]

### Question 4
In EX07, we have worked together to write a code to read the file earthquake.csv and assign each column to a list. We now want to further examine this earthquake catalog.

The code is shown below

In [20]:
import numpy as np
fp=open("earthquake.csv","r")
lines=fp.readlines()
fp.close()
year=[];month=[];day=[];time=[];mag=[];lon=[];lat=[];depth=[];region=[]
for line in lines[1:]:
    if "\"" in line:
        temp=line.split(",")
        year.append(temp[0]);month.append(temp[1])
        day.append(temp[2]);time.append(temp[3])
        mag.append(temp[4]);lon.append(temp[5]);lat.append(temp[6])
        depth.append(temp[7]);region.append(temp[8]+","+temp[9])
    else:
        temp=line.split(",")
        year.append(temp[0]);month.append(temp[1])
        day.append(temp[2]);time.append(temp[3])
        mag.append(temp[4]);lon.append(temp[5]);lat.append(temp[6])
        depth.append(temp[7]);region.append(temp[8])

The code above will give you variables of 'year', 'month', 'day', 'time', 'mag', 'lon', 'lat', 'depth', and 'regoin'. All these variables have `list` data type.
__Be careful that  Each element within the lists has `string` data type.__ Therefore, you will need to first convert strings to numbers before apply mathmatic operations on it.

In [2]:
# To convert a string to a number, we use float() or int() functions
a="10"
b=float(a)
c=int(a)
print(a,b,c)

10 10.0 10


(a) Use `for` loop, or list comprehension to convert all elements in the lists of 'mag','lon','lat',and 'depth' to floating numbers, and all elements in 'year', 'month', and 'day' to integer numbers. Remember, only change the data type of each element, and do not change the variable names.
* (10 points)

In [21]:
mag = [float(m) for m in mag]
lon = [float(l) for l in lon]
lat = [float(la) for la in lat]
depth = [float(d) for d in depth]
year = [int(y) for y in year]
month = [int(mo) for mo in month]
day = [int(da) for da in day]


(b) Using `for` loop to calculate the average magnitude of the earthquakes in this catalog.
* (10 points)

In [22]:
sum_mag = 0
for i in range(0,len(mag)):
    sum_mag = sum_mag + mag[i]
av_mag = sum_mag / len(mag)
av_mag

1.7830000000000006

(c) Calculate the average depth of the earthquakes whose magnitude is larger than 2.0
* (15 points)

In [24]:
sum_depth_2plus = 0
for i in range(0,len(mag)):
    if mag[i] > 2.0:
        sum_depth_2plus = sum_depth_2plus + depth[i]
av_depth_2plus = sum_depth_2plus / len(depth)
av_depth_2plus


12.037899999999995

(d) Use loop and condition to find the maximum depth of the earthquakes
* (20 points)

In [37]:
maxd = depth[0]
for i in range(1,len(depth)):
    if depth[i]>maxd:
        maxd = depth[i]
        imax = i
maxd

25

(e) In which region(s) does the largest earthquake(s) happen?
* (25 points)

In [46]:
max_mags = []
imax_ms = []
largest = mag[0]
for i in range(1,len(mag)):
    if mag[i] > largest:
        largest = mag[i]
        imag_max=i
ilargest = []
for i in range(0,len(mag)):
    if mag[i]==largest:
        ilargest.append(i)
if len(ilargest) == 1:
    ilargest = ilargest[0]
else:
    print('multiple earthquakes at max of %.1f mag'%largest)
    
region_of_largest = region[ilargest]
region_of_largest
        

        

'"MINDANAO, PHILIPPINES"'

(f) In which region(s) and at what year, month, day and time do the deepest earthquake(s) happen?
* (30 points)

In [50]:
deepest = []
ideepest = []
for i in range(0,len(depth)):
    if depth[i]==maxd:
        deepest.append(depth[i])
        ideepest.append(i)
if len(ideepest) == 1:
    ideepest = ideepest[0]
else:
    print('multiple earthquakes at depth of %.1f m'%deepest)
r_deepest = region[ideepest];y_deepest = year[ideepest]
m_deepest = month[ideepest];d_deepest = day[ideepest]
t_deepest = time[ideepest]
print('Information on the deepest earthquake:')
print('region:',r_deepest)
print('year:',y_deepest)
print('month:',m_deepest)
print('day:',d_deepest)
print('time:',t_deepest)

Information on the deepest earthquake:
region: FLORES SEA
year: 2019
month: 2
day: 10
time: 22:54:37
